app/controllers/saml_idp_controller.rb

```diff
@@ -94,27 +94,23 @@ def process_invalid_authn_context
     render nothing: true, status: :bad_request
   end
 
-  def authn_context_node
-    saml_request.document.xpath(
-      '//samlp:AuthnRequest/samlp:RequestedAuthnContext/saml:AuthnContextClassRef',
-      samlp: Saml::XML::Namespaces::PROTOCOL,
-      saml: Saml::XML::Namespaces::ASSERTION)
-  end
-
   def requested_authn_context
-    return authn_context_node[0].content if authn_context_node.length == 1
-
-    logger.info 'authn_context is missing'
-    nil
+    if saml_request.requested_authn_context
+      saml_request.requested_authn_context
+    else
+      logger.info 'authn_context is missing'
+      nil
+    end
   end
 
   def logout_response_builder
     SamlIdp::LogoutResponseBuilder.new(
       get_saml_response_id,
       issuer_uri,
-      saml_response_url,
+      saml_request.response_url,
       saml_request.request_id,
-      signature_opts)
+      SamlIdp.config.algorithm
+    )
   end
```

Comment:
This method is actually duplicated in the logout concern. We can get rid of the one here and keep the one in the concern.